### DISCUSSION TITLE
2024/03/03/why-array-index-begin-with-zero

### DISCUSSION THREAD

**Discussion by giscus** [kaochenlong/kaochenlong-discuss#5] [General]
# 2024/03/03/why-array-index-begin-with-zero

有沒有想過為什麼我們學的數學是從 1 開始算，但好像大部份比較主流的程式語言的陣列包括 Python 的串列，索引值卻是從 0 開始的？因為這是因為陣列的索引值並不是指向元素本身，比較像是記憶體位置的「偏移量」。假設有一個陣列像這樣：chars = ['a', 'b', 'c', 'd']當我們說「變數 char...

https://kaochenlong.com/2024/03/03/why-array-index-begin-with-zero.html

**Comment by JamesKai** [comment]
謝謝分享。 關於1-based index 我自己的理解是這樣設計比較符合傳統數學表達方式，尤其是矩陣的行列表達方法通常在文獻上中都是1-based index (我自己是沒有在數學課上聽過教授說0行0列這樣的話)。因此，文章中所說的那些語言基本上都是以處理數值計算 (包含向量、矩陣等) 見長的語言。其實還有不少語言也是從1開始，包含Matlab, Maple, Mathematica, Octave，全部都是以數學計算為主的語言。對這領域的人來說，由1開始確實是比較直覺。1-based index 比較少見大概是因為這些都是比較偏向科研人員使用的程式語言而非general purpose，所以受眾確實少很多。
